Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -133,34 +141,63 @@ internal void ValidateSecurePassword(SecureString password)
         private string ExtractPassword(SecureString password) =>
             password == null ? string.Empty : SecureStringHelper.Decode(password);
 
-        internal SFSession GetSession(string connStr, SecureString password)
+        internal SFSession GetSession(string connStr, SecureString password, SecureString passcode)
         {
             s_logger.Debug("SessionPool::GetSession" + PoolIdentification());
+            SFSession session = null;
+            var sessionProperties = SFSessionProperties.ParseConnectionString(connStr, password);
+            ValidatePoolingIfPasscodeProvided(sessionProperties);
             if (!GetPooling())
-                return NewNonPoolingSession(connStr, password);
+                return NewNonPoolingSession(connStr, password, passcode);
             var sessionOrCreateTokens = GetIdleSession(connStr);
+            if(sessionProperties.TryGetValue(SFSessionProperty.AUTHENTICATOR, out var authenticator) && authenticator == MFACacheAuthenticator.AUTH_NAME)
+                session = sessionOrCreateTokens.Session ?? NewSession(connStr, password, passcode, sessionOrCreateTokens.SessionCreationToken());
             if (sessionOrCreateTokens.Session != null)
             {
                 _sessionPoolEventHandler.OnSessionProvided(this);
             }
-            ScheduleNewIdleSessions(connStr, password, sessionOrCreateTokens.BackgroundSessionCreationTokens());
+            ScheduleNewIdleSessions(connStr, password, RegisterSessionCreationsToEnsureMinPoolSize());
```

Comment:
it is getting complicated so some refoctoring would be needed.
But I see the problem that the previous command `var sessionOrCreateTokens = GetIdleSession(connStr);` could generate some creation tokens. Ignoring them is not good because they are counted to the pool size.
You do `RegisterSessionCreationsToEnsureMinPoolSize()` but the previously generated tokens exists.
I belive GetIdleSession method should get a parameter which would be used to make a decisions if we are allowed to generate more creation tokens.

Reply:
The logic in the GetIdleSession was changed, so it only create the session token for the requested session, see line 249. So now the ownership of request the tokens for the additional session is using the RegisterSessionCreationsToEnsureMinPoolSize() method.